File: scaffold/src/main/java/top/xuqingquan/web/nokernel/WebUtils.java
package top.xuqingquan.web.nokernel;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.os.Handler;
import android.os.Looper;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.core.app.AppOpsManagerCompat;
import androidx.core.content.ContextCompat;
import androidx.core.os.EnvironmentCompat;
import top.xuqingquan.R;
import top.xuqingquan.utils.FileUtils;
import top.xuqingquan.utils.Timber;
import top.xuqingquan.web.publics.AbsAgentWebUIController;
import top.xuqingquan.web.publics.WebParentLayout;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.Method;
import java.text.SimpleDateFormat;
import java.util.*;

import static top.xuqingquan.web.nokernel.WebConfig.*;

/**
 * Created by 许清泉 on 2019-06-19 21:55
 */
public class WebUtils {
    private WebUtils(){}

    /**
     * @param context
     * @return WebView 的缓存路径
     */
    public static String getCachePath(Context context) {
        return context.getCacheDir().getAbsolutePath() + AGENTWEB_CACHE_PATCH;
    }

    public static String getAgentWebFilePath(Context context) {
        if (!TextUtils.isEmpty(AGENTWEB_FILE_PATH)) {
            return AGENTWEB_FILE_PATH;
        }
        String dir = getDiskExternalCacheDir(context);
        File mFile = new File(dir, FILE_CACHE_PATH);
        try {
            if (!mFile.exists()) {
                mFile.mkdirs();
            }
        } catch (Throwable throwable) {
            Timber.i("create dir exception");
        }
        Timber.i("path:" + mFile.getAbsolutePath() + "  path:" + mFile.getPath());
        AGENTWEB_FILE_PATH = mFile.getAbsolutePath();
        return AGENTWEB_FILE_PATH;
    }

    static String getDiskExternalCacheDir(Context context) {
        File mFile = context.getExternalCacheDir();
        if (mFile != null && Environment.MEDIA_MOUNTED.equals(EnvironmentCompat.getStorageState(mFile))) {
            return mFile.getAbsolutePath();
        }
        return null;
    }

    @SuppressWarnings("ResultOfMethodCallIgnored")
    public static File createFileByName(Context context, String name, boolean cover) throws IOException {
        String path = getAgentWebFilePath(context);
        if (TextUtils.isEmpty(path)) {
            return null;
        }
        File mFile = new File(path, name);
        if (mFile.exists()) {
            if (cover) {
                mFile.delete();
                mFile.createNewFile();
            }
        } else {
            mFile.createNewFile();
        }
        return mFile;
    }

    public static File createImageFile(Context context) {
        File mFile = null;
        try {
            String timeStamp =
                    new SimpleDateFormat("yyyyMMddHHmmss", Locale.getDefault()).format(new Date());
            String imageName = String.format("aw_%s.jpg", timeStamp);
            mFile = createFileByName(context, imageName, true);
        } catch (Throwable e) {
            e.printStackTrace();
        }
        return mFile;
    }

    public static Intent getIntentCaptureCompat(Context context, File file) {
        Intent mIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        Uri mUri = FileUtils.getUriFromFile(context, file);
        mIntent.addCategory(Intent.CATEGORY_DEFAULT);
        mIntent.putExtra(MediaStore.EXTRA_OUTPUT, mUri);
        return mIntent;
    }

    public static boolean isUIThread() {
        return Looper.myLooper() == Looper.getMainLooper();
    }

    private static Toast mToast = null;

    public static void toastShowShort(Context context, String msg) {
        if (mToast == null) {
            mToast = Toast.makeText(context.getApplicationContext(), msg, Toast.LENGTH_SHORT);
        } else {
            mToast.setText(msg);
        }
        mToast.show();
    }
    @Deprecated
    static void getUIControllerAndShowMessage(Activity activity, String message, String from) {
        if (activity == null || activity.isFinishing()) {
            return;
        }
        WebParentLayout mWebParentLayout = activity.findViewById(R.id.web_parent_layout_id);
        AbsAgentWebUIController mAgentWebUIController = mWebParentLayout.provide();
        if (mAgentWebUIController != null) {
            mAgentWebUIController.onShowMessage(message, from);
        }
    }

    public static boolean hasPermission(@NonNull Context context, @NonNull String... permissions) {
        return hasPermission(context, Arrays.asList(permissions));
    }

    public static boolean hasPermission(@NonNull Context context, @NonNull List<String> permissions) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return true;
        }
        for (String permission : permissions) {
            int result = ContextCompat.checkSelfPermission(context, permission);
            if (result == PackageManager.PERMISSION_DENIED) {
                return false;
            }
            String op = AppOpsManagerCompat.permissionToOp(permission);
            if (TextUtils.isEmpty(op)) {
                continue;
            }
            result = AppOpsManagerCompat.noteProxyOp(context, op, context.getPackageName());
            if (result != AppOpsManagerCompat.MODE_ALLOWED) {
                return false;
            }
        }
        return true;
    }

    public static List<String> getDeniedPermissions(Activity activity, String[] permissions) {
        if (permissions == null || permissions.length == 0) {
            return null;
        }
        List<String> deniedPermissions = new ArrayList<>();
        for (int i = 0; i < permissions.length; i++) {
            if (!hasPermission(activity, permissions[i])) {
                deniedPermissions.add(permissions[i]);
            }
        }
        return deniedPermissions;
    }

    private static Handler mHandler = null;

    public static void runInUiThread(Runnable runnable) {
        if (mHandler == null) {
            mHandler = new Handler(Looper.getMainLooper());
        }
        mHandler.post(runnable);
    }

    public static Method isExistMethod(Object o, String methodName, Class... clazzs) {
        if (null == o) {
            return null;
        }
        try {
            Class clazz = o.getClass();
            Method mMethod = clazz.getDeclaredMethod(methodName, clazzs);
            mMethod.setAccessible(true);
            return mMethod;
        } catch (Throwable throwable) {
            Timber.e(throwable);
        }
        return null;

    }
}

File: scaffold_web/src/main/java/top/xuqingquan/web/nokernel/WebUtils.kt
package top.xuqingquan.web.nokernel

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.os.Handler
import android.os.Looper
import android.provider.MediaStore
import android.support.v4.os.EnvironmentCompat
import android.text.TextUtils
import android.widget.Toast
import top.xuqingquan.utils.Timber
import top.xuqingquan.utils.getMIMEType
import top.xuqingquan.web.R
import top.xuqingquan.web.nokernel.WebConfig.AGENTWEB_CACHE_PATCH
import top.xuqingquan.web.nokernel.WebConfig.AGENTWEB_FILE_PATH
import top.xuqingquan.web.nokernel.WebConfig.FILE_CACHE_PATH
import top.xuqingquan.web.publics.WebParentLayout
import top.xuqingquan.web.provider.ScaffoldWebFileProvider
import java.io.File
import java.io.IOException
import java.lang.reflect.Method
import java.text.SimpleDateFormat
import java.util.*

/**
 * Created by 许清泉 on 2019-06-19 21:55
 */
object WebUtils {

    private var mToast: Toast? = null

    private var mHandler: Handler? = null

    /**
     * @param context
     * @return WebView 的缓存路径
     */
    @JvmStatic
    fun getCachePath(context: Context): String {
        return context.cacheDir.absolutePath + AGENTWEB_CACHE_PATCH
    }

    @JvmStatic
    fun getAgentWebFilePath(context: Context): String? {
        if (!TextUtils.isEmpty(AGENTWEB_FILE_PATH)) {
            return AGENTWEB_FILE_PATH
        }
        val dir = getDiskExternalCacheDir(context)
        val mFile = File(dir, FILE_CACHE_PATH)
        try {
            if (!mFile.exists()) {

                mFile.mkdirs()
            }
        } catch (throwable: Throwable) {
            Timber.i("create dir exception")
        }

        Timber.i("path:" + mFile.absolutePath + "  path:" + mFile.path)
        AGENTWEB_FILE_PATH = mFile.absolutePath
        return AGENTWEB_FILE_PATH
    }

    @JvmStatic
    internal fun getDiskExternalCacheDir(context: Context): String? {
        val mFile = context.externalCacheDir
        return if (mFile != null && Environment.MEDIA_MOUNTED == EnvironmentCompat.getStorageState(mFile)) {
            mFile.absolutePath
        } else null
    }

    @JvmStatic
    @Throws(IOException::class)
    fun createFileByName(context: Context, name: String, cover: Boolean): File? {
        val path = getAgentWebFilePath(context)
        if (path.isNullOrEmpty()) {
            return null
        }
        val mFile = File(path, name)
        if (mFile.exists()) {
            if (cover) {
                mFile.delete()
                mFile.createNewFile()
            }
        } else {
            mFile.createNewFile()
        }
        return mFile
    }

    @JvmStatic
    fun createImageFile(context: Context): File? {
        var mFile: File? = null
        try {
            val timeStamp = SimpleDateFormat("yyyyMMddHHmmss", Locale.getDefault()).format(Date())
            val imageName = "aw_$timeStamp.jpg"
            mFile = createFileByName(context, imageName, true)
        } catch (e: Throwable) {
            e.printStackTrace()
        }

        return mFile
    }

    @JvmStatic
    fun createVideoFile(context: Context): File? {
        var mFile: File? = null
        try {
            val timeStamp: String? =
                SimpleDateFormat("yyyyMMddHHmmss", Locale.getDefault())
                    .format(Date())
            val imageName = "aw_$timeStamp.mp4"  //默认生成mp4
            mFile = createFileByName(context, imageName, true)
        } catch (e: Throwable) {
            e.printStackTrace()
        }
        return mFile
    }

    @JvmStatic
    fun getIntentCaptureCompat(context: Context, file: File): Intent {
        val mIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        val mUri = getUriFromFile(context, file)
        mIntent.addCategory(Intent.CATEGORY_DEFAULT)
        mIntent.putExtra(MediaStore.EXTRA_OUTPUT, mUri)
        return mIntent
    }

    @JvmStatic
    fun getIntentVideoCompat(context: Context, file: File): Intent {
        val mIntent = Intent(MediaStore.ACTION_VIDEO_CAPTURE)
        val mUri = getUriFromFile(context, file)
        mIntent.addCategory(Intent.CATEGORY_DEFAULT)
        mIntent.putExtra(MediaStore.EXTRA_OUTPUT, mUri)
        return mIntent
    }

    @JvmStatic
    fun getCommonFileIntentCompat(context: Context, file: File): Intent {
        val mIntent = Intent().setAction(Intent.ACTION_VIEW)
        setIntentDataAndType(context, mIntent, getMIMEType(file), file, false)
        return mIntent
    }

    private fun setIntentDataAndType(
        context: Context,
        intent: Intent,
        type: String,
        file: File,
        writeAble: Boolean
    ) {
        if (Build.VERSION.SDK_INT >= 24) {
            intent.setDataAndType(getUriFromFile(context, file), type)
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            if (writeAble) {
                intent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
            }
        } else {
            intent.setDataAndType(Uri.fromFile(file), type)
        }
    }

    @SuppressLint("ShowToast")
    @JvmStatic
    fun toastShowShort(context: Context, msg: String) {
        if (mToast == null) {
            mToast = Toast.makeText(context.applicationContext, msg, Toast.LENGTH_SHORT)
        } else {
            mToast!!.setText(msg)
        }
        mToast!!.show()
    }

    @JvmStatic
    @Deprecated("")
    internal fun getUIControllerAndShowMessage(activity: Activity?, message: String, from: String) {
        if (activity == null || activity.isFinishing) {
            return
        }
        val mWebParentLayout = activity.findViewById<WebParentLayout>(R.id.scaffold_web_parent_layout_id)
        val mAgentWebUIController = mWebParentLayout.provide()
        mAgentWebUIController?.onShowMessage(message, from)
    }

    @JvmStatic
    fun isUIThread(): Boolean {
        return Looper.myLooper() == Looper.getMainLooper()
    }

    @JvmStatic
    fun runInUiThread(runnable: Runnable) {
        if (mHandler == null) {
            mHandler = Handler(Looper.getMainLooper())
        }
        mHandler!!.post(runnable)
    }

    @JvmStatic
    fun isExistMethod(o: Any?, methodName: String, vararg clazzs: Class<*>): Method? {
        if (null == o) {
            return null
        }
        try {
            val clazz = o.javaClass
            val mMethod = clazz.getDeclaredMethod(methodName, *clazzs)
            mMethod.isAccessible = true
            return mMethod
        } catch (throwable: Throwable) {
            Timber.e(throwable)
        }
        return null
    }

    @JvmStatic
    fun getUriFromFile(context: Context, file: File): Uri {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            ScaffoldWebFileProvider.getUriForFile(context, context.packageName + ".ScaffoldWebFileProvider", file)
        } else {
            Uri.fromFile(file)
        }
    }
}
